File: pca_lite/agents/writer.py
"""Writer Agent - write structured literature review drafts."""
from pca_lite.core.models import LiteratureEntry
from pca_lite.llm.base import BaseProvider


class WriterAgent:
    def __init__(self, llm_provider: BaseProvider):
        self.llm = llm_provider

    async def write_draft(
        self,
        themes: list[dict],
        literature_pool: list[LiteratureEntry],
        topics: list[str],
    ) -> dict:
        """Write a literature review draft from theme analysis.

        Args:
            themes: Theme analysis from Analyst Agent.
            literature_pool: Full literature pool for citation.
            topics: Topic keywords from config.

        Returns:
            Draft report dict with markdown content.
        """
        pool_indexed = self._build_indexed_pool(literature_pool)
        prompt = self._build_draft_prompt(themes, pool_indexed, topics)

        try:
            messages = [{"role": "user", "content": prompt}]
            draft_md = await self.llm.chat(messages)

            return {
                "agent": "writer",
                "action": "draft",
                "content": draft_md,
                "status": "draft",
            }
        except Exception as e:
            return {
                "agent": "writer",
                "action": "draft",
                "content": "",
                "status": "failed",
                "error": str(e),
            }

    def _build_indexed_pool(self, pool: list[LiteratureEntry]) -> str:
        return "\n".join(
            f"[{i + 1}] {e.title} | {e.year or 'n.d.'} | "
            f"{', '.join(e.authors[:3]) if e.authors else 'Unknown'} "
            f"| abstract: {e.abstract[:100] if e.abstract else 'N/A'}..."
            for i, e in enumerate(pool)
        )

    def _build_draft_prompt(
        self, themes: list[dict], pool: str, topics: list[str]
    ) -> str:
        theme_summaries = []
        for t in themes:
            paper_list = ", ".join(f"[{p}]" for p in t.get("paper_indices", []))
            theme_summaries.append(
                f"- Theme: {t.get('name', 'Unknown')}\n"
                f"  Description: {t.get('description', '')}\n"
                f"  Papers: {paper_list}\n"
                f"  Key Findings: {', '.join(t.get('key_findings', [])[:3])}\n"
                f"  Consensus: {t.get('consensus', 'unknown')}"
            )
        themes_text = "\n".join(theme_summaries)
        topics_text = ", ".join(topics[:5])

        return (
            f"Write a structured literature review on the following topics: {topics_text}\n\n"
            f"Literature Pool (use [index] to cite):\n{pool}\n\n"
            f"Theme Analysis:\n{themes_text}\n\n"
            "Write a literature review in Markdown format with sections: "
            "Abstract, 1. Introduction, 2. Methodology, 3. Findings (per theme), "
            "4. Discussion, 5. Conclusion. "
            "Use [index] citations only for papers in the literature pool above. "
            "Do not invent citations outside this pool."
        )

    async def run(
        self,
        themes: list[dict],
        literature_pool: list[LiteratureEntry],
        topics: list[str],
    ) -> dict:
        """Full writer workflow.

        Args:
            themes: Theme analysis from Analyst Agent.
            literature_pool: Full literature pool.
            topics: Topic keywords.

        Returns:
            Draft report.
        """
        draft = await self.write_draft(themes, literature_pool, topics)
        citation_check = self._check_citations(draft.get("content", ""), len(literature_pool))
        draft["citation_check"] = citation_check
        return draft

    def _check_citations(self, content: str, pool_size: int) -> dict:
        import re

        citation_pattern = re.compile(r"\[(\d+(?:-\d+)?)\]")
        citations = citation_pattern.findall(content)
        invalid = []
        for c in citations:
            if "-" in c:
                parts = c.split("-")
                try:
                    start, end = int(parts[0]), int(parts[1])
                    for idx in range(start, end + 1):
                        if idx < 1 or idx > pool_size:
                            invalid.append(f"{idx}")
                except ValueError:
                    invalid.append(c)
            else:
                try:
                    idx = int(c)
                    if idx < 1 or idx > pool_size:
                        invalid.append(c)
                except ValueError:
                    invalid.append(c)

        return {
            "valid": len(invalid) == 0,
            "invalid_citations": invalid,
        }

File: src-tauri/src/agents/writer.rs
use std::sync::Arc;
use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use crate::domain::agent::{Agent, AgentError};
use crate::domain::execution_context::TaskExecutionContext;
use crate::application::AppContainer;
use citeforge_core::entity::{LiteratureEntry, Theme};

#[derive(Debug, Serialize, Deserialize)]
pub struct WriterInput {
    pub task_id: String,
    pub topic: String,
    pub literature_entries: Vec<LiteratureEntry>,
    pub themes: Vec<Theme>,
    pub trends: Vec<String>,
    pub gaps: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WriterOutput {
    pub draft: String,
    pub citation_count: usize,
}

pub struct WriterAgent {
    container: Arc<AppContainer>,
}

impl WriterAgent {
    pub fn new(container: Arc<AppContainer>) -> Self {
        Self { container }
    }

    fn validate_citations(&self, draft: &str, pool_size: usize) -> Result<(), AgentError> {
        // Check for [N] format citations
        let citation_regex = regex::Regex::new(r"\[(\d+)\]").expect("citation regex should always be valid");
        let mut errors = Vec::new();

        for cap in citation_regex.captures_iter(draft) {
            if let Ok(idx) = cap[1].parse::<usize>() {
                if idx < 1 || idx > pool_size {
                    errors.push(format!("Invalid citation [{}] - pool size is {}", idx, pool_size));
                }
            }
        }

        if !errors.is_empty() {
            return Err(AgentError::Execution(errors.join("; ")));
        }

        Ok(())
    }
}

#[async_trait]
impl Agent for WriterAgent {
    type Input = WriterInput;
    type Output = WriterOutput;

    fn name(&self) -> &'static str {
        "WriterAgent"
    }

    async fn run(&self, ctx: &TaskExecutionContext, input: Self::Input) -> Result<Self::Output, AgentError> {
        tracing::info!("WriterAgent: writing draft for topic '{}'", input.topic);

        let pool_size = input.literature_entries.len();

        // Build literature context
        let literature_context: String = input.literature_entries.iter()
            .enumerate()
            .map(|(i, e)| format!("[{}] {} - {}", i + 1, e.title, e.authors.join(", ")))
            .collect::<Vec<_>>()
            .join("\n");

        let prompt = format!(
            "Write a structured literature review on: {}\n\n\
            Literature Pool:\n{}\n\n\
            Themes identified: {}\n\
            Trends: {}\n\
            Research gaps: {}\n\n\
            Write in sections: Abstract, Introduction, Methodology, Findings, Discussion, Conclusion.\n\
            Use [N] format for citations where N is the index in the literature pool above.\n\
            Ensure all citations are valid (1-based indexing into the literature pool).",
            input.topic,
            literature_context,
            input.themes.iter().map(|t| t.name.as_str()).collect::<Vec<_>>().join(", "),
            input.trends.join("; "),
            input.gaps.join("; ")
        );

        let messages = vec![
            citeforge_core::ports::ChatMessage {
                role: "user".to_string(),
                content: prompt,
            }
        ];

        let draft = ctx.run_with_timeout(async {
            self.container.llm.chat(messages).await
        }).await
        .map_err(|_| AgentError::Timeout)?
        .map_err(|e| AgentError::Llm(format!("LLM error: {}", e)))?;

        // Validate citations
        self.validate_citations(&draft, pool_size)?;

        // Count citations
        let citation_regex = regex::Regex::new(r"\[(\d+)\]").expect("citation regex should always be valid");
        let citation_count = citation_regex.find_iter(&draft).count();

        Ok(WriterOutput { draft, citation_count })
    }
}